From 7deec79beb60df7ecf031d449c99860741c4e889 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <jouni@qca.qualcomm.com>
Date: Fri, 30 Mar 2012 15:50:33 +0300
Subject: [PATCH 64/75] P2P: Use P2P Device ID attribute if Device Info not
 available

The "BSS p2p_dev_addr=address" command uses p2p_parse_dev_addr() to
figure out the P2P Device Address of the GO from scan results. This used
to work only if the P2P IE was received from Probe Response frames since
only those include the P2P Device Info attribute. Make this work with
Beacon frames, too, by using P2P Device ID attribute if the P2P Device
Info attribute is not present.

Signed-hostap: Jouni Malinen <jouni@qca.qualcomm.com>
(cherry picked from commit 526ec4aee84ed1ab25930d4445161a02aa5937cb)
---
 src/p2p/p2p.c |   13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

--- a/src/p2p/p2p.c
+++ b/src/p2p/p2p.c
@@ -2110,6 +2110,7 @@ int p2p_parse_dev_addr(const u8 *ies, si
 {
 	struct wpabuf *p2p_ie;
 	struct p2p_message msg;
+	int ret = -1;
 
 	p2p_ie = ieee802_11_vendor_ie_concat(ies, ies_len,
 					     P2P_IE_VENDOR_TYPE);
@@ -2121,14 +2122,16 @@ int p2p_parse_dev_addr(const u8 *ies, si
 		return -1;
 	}
 
-	if (msg.p2p_device_addr == NULL) {
-		wpabuf_free(p2p_ie);
-		return -1;
+	if (msg.p2p_device_addr) {
+		os_memcpy(dev_addr, msg.p2p_device_addr, ETH_ALEN);
+		ret = 0;
+	} else if (msg.device_id) {
+		os_memcpy(dev_addr, msg.device_id, ETH_ALEN);
+		ret = 0;
 	}
 
-	os_memcpy(dev_addr, msg.p2p_device_addr, ETH_ALEN);
 	wpabuf_free(p2p_ie);
-	return 0;
+	return ret;
 }
 
 
